Library/Homebrew/update_migrator/cache_entries_to_symlinks.rb

```diff
@@ -0,0 +1,93 @@
+require "hbc/cask_loader"
+require "hbc/download"
+
+module UpdateMigrator
+  module_function
+
+  def migrate_cache_entries_to_symlinks(initial_version)
+    return if initial_version && initial_version > "1.7.2"
+
+    return if ENV.key?("HOMEBREW_DISABLE_LOAD_FORMULA")
+
+    ohai "Migrating cache entries..."
+
+    load_formula = lambda do |formula|
+      begin
+        Formula[formula]
+      rescue FormulaUnavailableError
+        nil
+      end
+    end
+
+    load_cask = lambda do |cask|
+      begin
+        Hbc::CaskLoader.load(cask)
+      rescue Hbc::CaskUnavailableError
+        nil
+      end
+    end
+
+    formula_downloaders = if HOMEBREW_CACHE.directory?
+      HOMEBREW_CACHE.children
+                    .reject(&:symlink?)
+                    .select(&:file?)
+                    .map { |child| child.basename.to_s.sub(/\-\-.*/, "") }
+                    .uniq
+                    .map(&load_formula)
+                    .compact
+                    .flat_map { |formula| formula_resources(formula) }
+                    .map { |resource| [resource.downloader, resource.download_name, resource.version] }
+    else
+      []
+    end
+
+    cask_downloaders = if (HOMEBREW_CACHE/"Cask").directory?
+      (HOMEBREW_CACHE/"Cask").children
+                             .reject(&:symlink?)
+                             .select(&:file?)
+                             .map { |child| child.basename.to_s.sub(/\-\-.*/, "") }
+                             .uniq
```

Comment:
Can the logic above maybe be shared between the above two variable definitions into a separate method?